src/Microsoft.ComponentDetection.Common/DependencyGraph/DependencyGraph.cs

```diff
@@ -141,6 +142,30 @@ public ICollection<string> GetAncestors(string componentId)
             .ToList();
     }
 
+    public HashSet<TypedComponent> GetAncestorsAsTypedComponents(string componentId, Func<string, TypedComponent> toTypedComponent)
+    {
+        ArgumentNullException.ThrowIfNull(componentId);
+        return this.GetAncestors(componentId)
+            .Select(a => this.componentNodes[a].TypedComponent ?? toTypedComponent(a))
+            .ToHashSet(new ComponentComparer());
+    }
+
+    public HashSet<TypedComponent> GetRootsAsTypedComponents(string componentId, Func<string, TypedComponent> toTypedComponent)
+    {
+        ArgumentNullException.ThrowIfNull(componentId);
+        return this.GetExplicitReferencedDependencyIds(componentId)
+            .Select(r => this.componentNodes[r].TypedComponent ?? toTypedComponent(r))
```

Comment:
Are we certain `this.componentNodes[r]` will exist, should we check first before `this.componentNodes[r].TypedComponent`?

Reply:
the `GetAncestors` and `GetExplicitReferencedDependencyIds` calls do confirm that all of the nodes that are returned are in the componentNodes list, but it definitely doesn't hurt to not make that assumption in case things change in the future. I'll add a quick check